### PR TITLE
Fix Multipart not to raise an exception when closing boundary only

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
A `Multipart` can be terminated with the ending boundary without a body part. For example, the following request is totally valid:
```
POST /test.html HTTP/1.1
Host: example.org
Content-Type: multipart/form-data;boundary="boundary"

--boundary--
```
We must not raise an exception in that case.

Modifications:
- Do not raise an exception when `Multipart` has the ending boundary only.
- Propagate an exception in `MultipartDecoder` to the subscriber correctly.
- Change to return `false` from `StreamMessage.isEmpty()` when the `StreamMessage` is created from a empty `Multipart`
  - Because it produces the ending boundary.

Result:
- You no longer see a 500 response when an empty multipart request is sent.

### CODE REVIEW

**Comment by chungonn** [comment]
@minwoox Thank You for working on this :)

File: core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java [hunk 1]
```diff
@@ -201,42 +202,30 @@ void parse() {
...
                         logger.trace("Skipped the preamble.");
-                        state = State.START_PART;
+                        state = State.START_PART_OR_END_MESSAGE;
                         break;
 
-                    case START_PART:
-                        logger.trace("state={}", State.START_PART);
+                    case START_PART_OR_END_MESSAGE:
+                        logger.trace("state={}", State.START_PART_OR_END_MESSAGE);
+                        final String headerLine = readHeaderLine();
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java:210]: As per https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html, `\r\n` could be appended to the closing boundary.
```
--simple boundary--\r\nThis is the epilogue.  It is also to be ignored.
```
Could we also consider the corner case?

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceMultipartTest.java [hunk 2]
```diff
@@ -78,6 +88,33 @@ void testUploadFile(String path) throws Exception {
                            "\"param1\":\"armeria\"}");
     }
 
+    @Test
+    void emptyBodyPart() {
+        final Multipart multipart = Multipart.of();
+        final HttpRequest request = multipart.toHttpRequest("/uploadWithMultipartObject");
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceMultipartTest.java:94]: Should we add a unit test for an empty body to `MimeParserTest`?

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceMultipartTest.java:94]: Thanks! added. 😉

File: core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java]: Thanks! fixed. 😉

File: core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java [hunk 3]
```diff
@@ -415,38 +415,78 @@ private static ByteBuf safeReadBytes(StreamDecoderInput in, int length) {
...
         }
 
-        // Check for \n or \r\n
-        if (boundaryStart + boundaryLength + linearWhiteSpace < length &&
-            (in.getByte(boundaryStart + boundaryLength + linearWhiteSpace) == '\n' ||
-             in.getByte(boundaryStart + boundaryLength + linearWhiteSpace) == '\r')) {
+        // Three valid cases:
+        // boundary+whitespace+"\n"
+        // boundary+whitespace+"\r\n"
+        // boundary+whitespace+"--" + "whatever after --"
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java:437]: Question: If the third case means a closing boundary, shouldn't `whitespace` be considered?
For example:
```suggestion
        // boundary+"--" + "whatever after --"
```

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java]: That's a good point. 👍

**Review by ikhoon — COMMENTED** [review]
Looking good. Left minor comments.

File: core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java [hunk 4]
```diff
@@ -415,38 +415,86 @@ private static ByteBuf safeReadBytes(StreamDecoderInput in, int length) {
...
 
-        // Consider all the whitespace boundary+whitespace+"\r\n"
+        // Three valid cases:
+        // boundary+"--" + "whatever after --" // closing boundary
+        // boundary+whitespace+"\n"
+        // boundary+whitespace+"\r\n"
+
+        // Check the closing boundary first.
+        int followingCharOffset = boundaryStartOffset + boundaryLength;
+        if (followingCharOffset >= length) {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java:434]: Sounds good. Let me change that.

File: core/src/main/java/com/linecorp/armeria/common/multipart/DefaultMultipart.java [hunk 5]
```diff
@@ -173,7 +173,7 @@ public boolean isOpen() {
 
     @Override
     public boolean isEmpty() {
-        return parts.isEmpty();
+        return false;
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/DefaultMultipart.java:176]: I left why I changed this on the commit message:
> Change to return false from StreamMessage.isEmpty() when the StreamMessage is created from a empty Multipart
Because it produces the ending boundary.

[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/DefaultMultipart.java:176]: Oops, it should only be `false` when `StreamMessage.isEmpty()` is called. Let me fix that. 😉

[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/multipart/DefaultMultipart.java:176]: This method is only called after a `Multipart` is converted into a `StreamMessage` and it's not always empty because it produces at least a closing boundary. So let me add a comment and keep that as is.

[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/multipart/DefaultMultipart.java:179]: Understood. 👍

**Review by ikhoon — APPROVED** [review]
Nice work, @minwoox! 💯

**Review by jrhee17 — APPROVED** [review]
Looks good to me! 👍  Thanks @minwoox 🙇 👍 🙇

**Comment by minwoox** [comment]
Thanks, reviewers. 😉
